Support RPC server and client functions

LiZhang-2020 opened the pull request:
jsonrpc: support RPC server and client functions

Support basic Json RPC functions,
such as RPC server and client.
Depends on libev http://software.schmorp.de/pkg/libev.html

examples/vdpa: support RPC message and handle

Create VDPA RPC thread to handle RPC message from App.

kailiangz1 commented on lib/jsonrpc/cJSON.c:
```diff
@@ -0,0 +1,1049 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
```
like this:
/* SPDX-License-Identifier: BSD-3-Clause
 * Copyright (c) 2022 NVIDIA Corporation & Affiliates
 */

LiZhang-2020 replied:
This file is coded by Henirgue from open source and copyright is not belong to NVIDIA.
Such as:
Copyright (c) 2012-2013 Henrique Gomes

kailiangz1 commented on examples/vdpa/vdpa_rpc.c:
```diff
@@ -0,0 +1,251 @@
+	pthread_mutex_lock(&rpc_ctx->rpc_lock);
+	if (pf_add && pf_dev) {
+		/* Parse PCI device name*/
+		RTE_LOG(ERR, RPC, "lizh cJSON mgmtpf.pf_add %s:%d pf_dev %s: %s...start \n", pf_add->string, pf_add->valueint, pf_dev->string, pf_dev->valuestring);
```
lizh should not here

LiZhang-2020 replied:
Will deleted all log with lizh.

kailiangz1 commented on examples/vdpa/vdpa_rpc.c:
```diff
@@ -0,0 +1,251 @@
+	pthread_mutex_lock(&rpc_ctx->rpc_lock);
+	if (pf_add && pf_dev) {
+		/* Parse PCI device name*/
+		RTE_LOG(ERR, RPC, "lizh cJSON mgmtpf.pf_add %s:%d pf_dev %s: %s...start \n", pf_add->string, pf_add->valueint, pf_dev->string, pf_dev->valuestring);
```
one line should not exceed 80 character

LiZhang-2020 replied:
Done

kailiangz1 commented on examples/vdpa/main.c:
```diff
@@ -529,6 +533,7 @@ main(int argc, char *argv[])
 	struct rte_device *dev;
 	struct rte_dev_iterator dev_iter;
 
+	vdpa_rpc_start(&vdpa_rpc_ctx);
```
judge return value

LiZhang-2020 replied:
Done

kailiangz1 commented on examples/vdpa/vdpa_rpc.c:
```diff
@@ -0,0 +1,251 @@
+	cJSON *result = NULL;
+	struct vdpa_rpc_context *rpc_ctx;
+
+	RTE_LOG(ERR, RPC, "lizh cJSON mgmtpf...start \n");
```
INFO

LiZhang-2020 replied:
Done

kailiangz1 commented on examples/vdpa/vdpa_rpc.c:
```diff
@@ -0,0 +1,251 @@
+	pthread_attr_init(&attr);
+	ret = pthread_attr_setschedpolicy(&attr, SCHED_RR);
+	if (ret) {
+		RTE_LOG(ERR, RPC, "Failed to set thread sched policy = RR.\n");
```
print ret value and return ret

LiZhang-2020 replied:
Done

kailiangz1 commented on examples/vdpa/vdpa_rpc.h:
```diff
@@ -0,0 +1,40 @@
+ * Copyright 2022, NVIDIA CORPORATION & AFFILIATES.
+ */
+
+#ifndef _VHOST_BLK_H_
```
_VDPA_RPC_H_

LiZhang-2020 replied:
Done

kailiangz1 commented on examples/vdpa/vdpa_rpc.h:
```diff
@@ -0,0 +1,40 @@
+#include <unistd.h>
+#include <sys/eventfd.h>
+
+#include <rte_vhost.h>
```
why need this 2 header file?
vhost.h eventfd.h

LiZhang-2020 replied:
Delete them

kailiangz1 commented on lib/jsonrpc/version.map:
```diff
@@ -9,6 +9,7 @@ DPDK_22 {
 	jrpc_register_procedure;
 	jrpc_deregister_procedure;
 	jsonrpc_client_call_method;
+	rte_rpc_vdpa_global_ops_init;
```
who will call rte_rpc_vdpa_global_ops_init?

LiZhang-2020 replied:
Admin queue will use it to register the PF driver ops functions.

kailiangz1 commented on examples/vdpa/main.c:
```diff
@@ -549,20 +577,6 @@ main(int argc, char *argv[])
 		cmdline_interact(cl);
 		cmdline_stdin_exit(cl);
 	} else {
-		RTE_DEV_FOREACH(dev, "class=vdpa", &dev_iter) {
```
should not remove prvious logic.

LiZhang-2020 replied:
Done

kailiangz1 commented on examples/vdpa/vdpa_rpc.c:
```diff
@@ -0,0 +1,358 @@
+		result = vdpa_pf_dev_remove(pf_dev->valuestring);
+	}
+	if (pf_list) {
+		result = vdpa_pf_dev_list();
```
else if{
else if
they can't true in the same time ?

LiZhang-2020 replied:
Done

GavinLi-NV commented on lib/jsonrpc/rte_rpc.h:
```diff
@@ -0,0 +1,101 @@
+	VDPA_VF_MSIX_NUM,
+	VDPA_VF_QUEUE_NUM,
+	VDPA_VF_QUEUE_SZIE,
+	VDPA_VF_NET_FEATURES,
```
Remove "_NET"

LiZhang-2020 replied:
Done

GavinLi-NV commented on examples/vdpa/main.c:
```diff
@@ -92,10 +96,10 @@ parse_args(int argc, char **argv)
 	}
 
-	if (iface[0] == '\0' && interactive == 0) {
+	/*if (iface[0] == '\0' && interactive == 0) {
```
Still temporary code?

LiZhang-2020 replied:
Delete it

GavinLi-NV commented on examples/vdpa/vdpa_rpc.h:
```diff
@@ -0,0 +1,40 @@
+#define VDPA_LOCAL_HOST "127.0.0.1"
+
+enum {
+	VDPA_RPC_PORT = 12190,
```
Could user change this in runtime?

LiZhang-2020 replied:
No, it is hardcode. It is exampler for customer. They can change it in his code

GavinLi-NV commented on lib/jsonrpc/version.map:
```diff
@@ -0,0 +1,15 @@
+DPDK_22 {
```
maybe we should use "INTERNAL" instead?

LiZhang-2020 replied:
Done

GavinLi-NV commented on lib/jsonrpc/rte_rpc.h:
```diff
@@ -0,0 +1,106 @@
+enum vdpa_vf_modify_flags {
+	VDPA_VF_MSIX_NUM,
+	VDPA_VF_QUEUE_NUM,
+	VDPA_VF_QUEUE_SZIE,
```
Typo

LiZhang-2020 replied:
Done

LiZhang-2025 committed:
app/vfe-vdpa: support RPC server and client functions

Add basic Json RPC functions, such as RPC server and client.
The following files is porting from open source
and author is hmg.
cJSON.c/cJSON.h/jsonrpc-c.c/jsonrpc-c.h
https://github.com/hmng/jsonrpc-c

These new files depend on libev,
http://software.schmorp.de/pkg/libev.html
It need install libev in system.

Signed-off-by: Li Zhang <lizh@nvidia.com>

LiZhang-2025 committed:
app/vfe-vdpa: support RPC message and handle

Support RPC message with application user
to configure VDPA.

Create RPC thread to handle RPC message from application.
Call the relation API rte_vdpa to trigger
driver configuration HW.

python file vhostmgmt as RPC example.
The following command as example:
python vhostmgmt -h
python vhostmgmt mgmtpf -h
python vhostmgmt mgmtpf -a 000:04:00.0
python vhostmgmt mgmtpf -r 000:04:00.0
python vhostmgmt mgmtpf -l
python vhostmgmt vf -p 000:04:00.3 -n 24 -q 256 -s 1024 -f 200000002 -t 2048 -m 00:11:22:33:44:55
python vhostmgmt vf -a 000:04:00.3 -v /tmp/sock1
python vhostmgmt vf -l 000:04:00.3
python vhostmgmt vf -i 000:04:00.3
python vhostmgmt vf -r 000:04:00.3
python vhostmgmt vf -h
python vhostmgmt vf -d 000:04:00.3 -o 1
python vhostmgmt vf -d 000:04:00.3 -o 4 -b 2 -e 4096 -g 4096

Signed-off-by: Li Zhang <lizh@nvidia.com>
